src/ansys/grantami/recordlists/_connection.py

```diff
@@ -48,38 +48,160 @@
+        """
+        status_codes = set()
+        for client_version, client in clients.items():
+            try:
+                server_version = client._get_mi_server_major_minor_version()
+                if server_version:
+                    return server_version
+
+            except KeyError:  # Deserialization error, raised if a model has changed name
+                pass
+
+            except requests.exceptions.RequestException:
+                raise ConnectionError(
+                    "An unexpected error occurred when trying to connect Granta MI Server API. Check "
+                    "that SSL certificates have been configured for communications between Granta MI "
+                    "Server and client Granta MI applications."
+                )
+
+            except ApiException as e:
+                status_codes.add(e.status_code)
+
+        if status_codes == {404}:
+            # No client version could get the version.
+            raise ConnectionError(
+                "Cannot find the Server API definition in Granta MI Service Layer. Ensure a compatible version of "
+                "Granta MI is available try again. This package supports Granta MI versions "
+                f"{self._all_mi_versions_formatted}."
+            )
+        else:
+            # At least one client version found the resource, but the version number was still not returned.
+            raise ConnectionError(
+                "An unexpected error occurred when trying to connect Server API in Granta MI Service Layer. Check "
+                "the Service Layer logs for more information, and that a compatible version of Granta MI is available "
+                f"and try again. This package supports Granta MI versions {self._all_mi_versions_formatted}."
+            )
+
+    @property
+    def _all_mi_versions(self) -> Iterable[tuple[int, int]]:
+        """All Granta MI versions supported by this package.
+
+        Returns
+        -------
+        list of 2-tuple of int
+            List of major.minor version numbers.
+        """
+        return self._client_map.keys()
+
+    @property
+    def _all_mi_versions_formatted(self) -> str:
+        """All Granta MI versions supported by this package, formatted as a string.
+
+        Returns
+        -------
+        str
+            A formatted string of all versions supported by this package.
+        """
+        formatted_versions = [
+            ".".join(str(e) for e in version) for version in self._all_mi_versions
+        ]
+        return ",".join(formatted_versions)
+
+    @property
+    def _highest_mi_version(self) -> tuple[int, int]:
+        """The highest Granta MI version supported by this package.
+
+        Returns
+        -------
+        2-tuple of int
+            major.minor version number.
+        """
+        result = (0, 0)
+        for version in self._client_map.keys():
+            if version > result:
```

Comment:
I'm surprised this works - does comparison for tuples sort by each element in turn? How clever!

Reply:
It does! It's almost like it was designed for comparing version numbers.